README.md
#DESCRIPTION
World of Warcraft addon for automatic EP/GP charging by combatlog events: deaths, damage, debuffs, interrupts and dispels. May be useful for saving your raid time due to the new encounters mastering.

You may add/remove/enable/disable rules via console commands.

#DEPENDENCIES

This addon works only with [epgp addon](http://www.curse.com/addons/wow/epgp-dkp-reloaded) installed.

#CONFIGURATION
Addon configuration may be managed via `/epgpbb` or `/ebb` slash commands:

* `/ebb help` - display configuration help 
* `/ebb status` - display addon settings
* `/ebb announce (say|guild|raid|party)` - announce active rules to a channel. Raid by default, or say if player is not in raid
* `/ebb list` - list of current configuration and rules
* `/ebb add value (GP|EP) for event` - add new rule to charge GP or EP on event. Negative values supported. Events are:
  * `death by spell_id` - taking damage with overkill from spell by id
  * `damagetaken by spell_id [min_damage]` - taking min_damage damage from spell by id, any damage by default
  * `buff by spell_id [min_stacks]` - gaining min_stacks of buff/debuff. 1 by default
  * `interrupt spell_id` - interrupting spell
  * `dispel spell_id` - dispelling/spellstealing spell
* `/ebb protect by` - protect from damage taken penalties of penalty_class by conditions:
  * `cast spell_id` - casting spell with spell_id (long important heals)
  * `buff spell_id` - having buff with spell_id (anti-magic shell)
* `/ebb del rule_id` - delete rule by id
* `/ebb enable rule_id` - enable rule by id
* `/ebb disable rule_id` - disable rule by id
* `/ebb reset` - reset configuration, delete all rules. Note: no confirmation asked
* `/ebb autologging (on|off)` - enables or disables automatic combatlogging on on/off command
* `/ebb on` - enables monitoring, start combat logging if autologging is enabled (put this in pull macro)
* `/ebb off` - disables monitoring, stops combat logging if autologging is disabled (put this in wipe macro)

#EXAMPLE RULES
My guilds Blackrock Foundry penalties (except last boss and Blast Furnance)

    /ebb add 150 GP for death by 156554       
    /ebb add 150 GP for buff by 154960        
    /ebb add 50 GP for damagetaken by 157247  
    /ebb add 50 GP for damagetaken by 158140  
    /ebb add 50 GP for damagetaken by 161570  
    /ebb add 150 GP for death by 156938       
    /ebb add 150 GP for death by 154938       
    /ebb add 50 GP for buff by 155314         
    /ebb add 150 GP for buff by 154989 3      
    /ebb add 150 GP for damagetaken by 160050 
    /ebb add 150 GP for damagetaken by 157659 
    /ebb add 150 GP for damagetaken by 161839 
    /ebb add 150 GP for damagetaken by 157884 
    /ebb add 150 GP for damagetaken by 160733 
    /ebb add 150 GP for damagetaken by 176133

#LINKS
* [Curse page](http://www.curse.com/addons/wow/epgp-battle-bot)
* [Repository](https://github.com/hurricup/WoW-EPGP-BattleBot)
* [Bug tracker](https://github.com/hurricup/WoW-EPGP-BattleBot/issues)

#AUTHOR
(c) 2015 Alexandr Evstigneev (hurricup@evstigneev.com)


CHANGES.md
#Release 6.1.4 (in development)
* Migrated to Ace libraries

#Release 6.1.3, Sun, 15 Mar 2015
* Added checking that rule target is a player, not a pet or something
* Fixed typo in help for `buff` rule
* Added optional `min_damage` argument for `damagetaken` rule, 0 by default

#Release 6.1.2, Sat, 14 Mar 2015
* Fixed bug with announcement channel selecting, thanks to @SimplEMeaT 

#Release 6.1.1, Sun, 08 Mar 2015
* Help is now being printed line by line
* Added `protect` command with `buff` and `cast` options. Protects from taking damage penalties
* Announce is now properly using say/raid channels by default depending on your raid status
* Implemented `autologging` command
* Implemented `status` command
* Implemented EP and GP support. Also, negative values may be specified
* Implemented `interrupt` event
* Implemented `dispel` event
* Syntax change: `on event` changed to `for event`

epgp_battle_bot/locale-en.lua
EPGP_BB_STATUS_VERSION = "WowRaider.Net EPGP Battle Bot %s."
EPGP_BB_STATUS_STATUS = "  Events monitoring: %s"
EPGP_BB_STATUS_AUTOLOGGING = "  Automatic combatlogging: %s"
EPGP_BB_STATUS_RULES = "  There are %s rules (type /ebb list to list them)"

EPGP_BB_HELP = {
    EPGP_BB_STATUS_VERSION, 
    "Configurable by /epgpbb or /ebb commands. Options list:",
    "  /ebb help - display configuration help",
    "  /ebb status - displays current addon settings",
    "  /ebb announce (say|guild|raid|party) - announce active rules to the specific channel (raid by default, say if not in raid)",
    "  /ebb list - display the list of current rules",
    "  /ebb add value (EP|GP) for event - add new rule to charge GP or EP on event. Negative values are supported. Events are:",
    "    death by spell_id - taking damage with overkill from spell by id",
    "    damagetaken by spell_id [min_damage] - taking min_damage damage from spell by id, any amount by default",
    "    buff by spell_id [min_staks] - gaining min_staks of buff/debuff, 1 by default",
    "    interrupt spell_id - interrupting specific spell",
    "    dispel spell_id - dispelling specific buff/debuff",
    "  /ebb protect by - protect from damage taken penalties of penalty_class by conditions:",
    "    cast spell_id - casting spell with spell_id (long important heals)",
    "    buff spell_id - having buff with spell_id (anti-magic shell)",
    "  /ebb del rule_id - delete rule by id",
    "  /ebb enable rule_id - enable rule by id",
    "  /ebb disable rule_id - disable rule by id",
    "  /ebb reset - reset configuraion, delete all rules",
    "  /ebb autologging (on|off) - enables or disables automatic combatlogging on on/off command",
    "  /ebb on - enables monitoring, start combat logging, put this in pull macro",
    "  /ebb off - disables monitoring, stops combat logging, put this in wipe macro",
}

EPGP_BB_CONFIG_RESET = "Configuration has been reset\n"
EPGP_BB_RULE_PH = "%d: (%s) %s" -- rule number, status, rule text

EPGP_BB_RULE_PROTECT_CAST_PH = "immune to damage penalties while casting %s"
EPGP_BB_RULE_PROTECT_BUFF_PH = "immune to damage penalties with buff %s"

EPGP_BB_RULE_INTERRUPT_PH = "%d %s for interrupting %s"
EPGP_BB_RULE_DISPEL_PH = "%d %s for dispelling/stealing %s"
EPGP_BB_RULE_DAMAGE_TAKEN_BY_PH = "%d %s for taking damage from %s"
EPGP_BB_RULE_DAMAGE_TAKEN_AMOUNT_BY_PH = "%d %s for taking more than %s damage from %s"
EPGP_BB_RULE_DEATH_BY_PH = "%d %s for death from %s"
EPGP_BB_RULE_BUFF_BY_PH = "%d %s for gaining buff %s"
EPGP_BB_RULE_BUFF_STACKS_BY_PH = "%d %s for gaining %d stacks of the buff %s"

EPGP_BB_ACTIVE_RULES_HEADER = "List of active penalties:"
EPGP_BB_RULE_NOT_FOUND = "Error: unable to find rule %s. Here is the list of available rules:"
EPGP_BB_RULE_ENABLED = "Enabled rule: %s"
EPGP_BB_RULE_DISABLED = "Disabled rule: %s"
EPGP_BB_RULE_DELETED = "Deleted rule: %s"
EPGP_BB_STATE = {
    [true] = "enabled",
    [false] = "disabled",
}
EPGP_BB_SPELL_LINK = "|cff71d5ff|Hspell:%d|h[%s]|h|r"
EPGP_BB_CREATED_RULE = "Created rule: %s"
EPGP_BB_REPLACED_RULE = "Replaced rule: '%s' with '%s'"
EPGP_BB_ADDON_ENABLED = "EPGP Battle Bot enabled, be careful"
EPGP_BB_ADDON_DISABLED = "EPGP Battle Bot disabled, relax"
EPGP_BB_LOGGING_ENABLED = "Combatlog recording turned on"
EPGP_BB_LOGGING_DISABLED = "Combatlog recording turned off"

EPGP_BB_AUTOLOGGING_ENABLED = "Automatic combatlogging control enabled"
EPGP_BB_AUTOLOGGING_DISABLED = "Automatic combatlogging control disabled"


epgp_battle_bot/locale-ru.lua
if GetLocale() ~= "ruRU" then 
    return 
end

EPGP_BB_STATUS_VERSION = "WowRaider.Net EPGP Battle Bot %s."
EPGP_BB_STATUS_STATUS = "  Отслеживание событий: %s"
EPGP_BB_STATUS_AUTOLOGGING = "  Автоматическая запись комбатлога: %s"
EPGP_BB_STATUS_RULES = "  Добавлено правил: %s (введите /ebb list для просмотра)"

EPGP_BB_HELP = {
    EPGP_BB_STATUS_VERSION,
    "Команды настройки /epgpbb или /ebb. Список опций:",
    "  /ebb help - показать помощь",
    "  /ebb status - показать текущие настройки аддона",
    "  /ebb announce (say|guild|raid|party) - анонсировать активный набоор правил в указанный канал (по умолчанию, в рейд или в say, если игрок не в рейде)",
    "  /ebb list - показать список существующих правил",
    "  /ebb add value (EP|GP) for event - добавить новое правило начисления EP или GP при событии. Поддерживаются отрицательные числа. Доступные события:",
    "    death by spell_id - получение избыточного урона от способности по ее id",
    "    damagetaken by spell_id [min_damage] - получение min_damage урона от способности по id. По умолчанию - любого урона",
    "    buff by spell_id [min_staks] - получение min_staks стаков баффа/дебаффа. 1 по умолчанию",
    "    interrupt spell_id - прерывание указанного заклинания",
    "    dispel spell_id - рассеивание баффа/дебаффа",
    "  /ebb protect by - защищает от штрафов за получение урона при условии:",
    "    cast spell_id - чтение заклинания spell_id (длинные и важные хилы)",
    "    buff spell_id - наличие баффа spell_id (антимагический панцирь)",
    "  /ebb del rule_id - удалить правило по id",
    "  /ebb enable rule_id - включить правило по id",
    "  /ebb disable rule_id - выключить правило по id",
    "  /ebb reset - сбросить настройки, удалть все правила",
    "  /ebb autologging (on|off) - включить или выключить автоматическую запись комбатлога при включении или выключении отслеживания",
    "  /ebb on - включить отслеживание, удобно использовать в макросе на пулл",
    "  /ebb off - отключает мониторинг, удобно использовать в макросе на вайп.",
}
EPGP_BB_CONFIG_RESET = "Конфигурация сброшена\n"
EPGP_BB_RULE_PH = "%d: (%s) %s" -- rule number, status, rule text
EPGP_BB_ACTIVE_RULES_HEADER = "Список действующих штрафов:"

EPGP_BB_RULE_DEATH_BY_PH = "%d %s за смерть от %s"
EPGP_BB_RULE_BUFF_BY_PH = "%d %s за получение баффа %s"
EPGP_BB_RULE_DAMAGE_TAKEN_BY_PH = "%d %s за получение урона от %s"
EPGP_BB_RULE_DAMAGE_TAKEN_AMOUNT_BY_PH = "%d %s за получение более %s урона от %s"
EPGP_BB_RULE_BUFF_STACKS_BY_PH = "%d %s за получение %d стаков бафа %s"
EPGP_BB_RULE_INTERRUPT_PH = "%d %s за прерывание %s"
EPGP_BB_RULE_DISPEL_PH = "%d %s за рассеивание/кражу %s"

EPGP_BB_RULE_PROTECT_CAST_PH = "иммунитет к штрафам за урон во время чтения заклинания %s"
EPGP_BB_RULE_PROTECT_BUFF_PH = "иммунитет к штрафам за урон с баффом %s"

EPGP_BB_RULE_NOT_FOUND = "Ошибка. Не могу найти правило %s. Список существующих правил:"
EPGP_BB_RULE_ENABLED = "Включено правило: %s"
EPGP_BB_RULE_DISABLED = "Выключено правило: %s"
EPGP_BB_RULE_DELETED = "Удалено правило: %s"
EPGP_BB_STATE = {
    [true] = "включено",
    [false] = "выключено",
}
EPGP_BB_SPELL_LINK = "|cff71d5ff|Hspell:%d|h[%s]|h|r"
EPGP_BB_CREATED_RULE = "Создано правило: %s"
EPGP_BB_REPLACED_RULE = "Замена правила: '%s' на '%s'"
EPGP_BB_ADDON_ENABLED = "EPGP Battle Bot активирован, будьте осторожны"
EPGP_BB_ADDON_DISABLED = "EPGP Battle Bot деактивирован, расслабьтесь"
EPGP_BB_LOGGING_ENABLED = "Включена запись журнала сражений"
EPGP_BB_LOGGING_DISABLED = "Запись журнала сражений выключена"

EPGP_BB_AUTOLOGGING_ENABLED = "Автоматическое управление записью журнала сражений включено"
EPGP_BB_AUTOLOGGING_DISABLED = "Автоматическое управление записью журнала сражений отключено"


epgp_battle_bot/battle_bot.lua
EBB = LibStub("AceAddon-3.0"):NewAddon("EPGPBattleBot", "AceConsole-3.0", "AceEvent-3.0")
local EBB = EBB
local GS = LibStub("LibGuildStorage-1.2")

---------------------- pre Ace section -----------------------------------------
-- legacy thing, remove with import_legacy_rules
local config_keys = {
    "death",
    "damagedone",
    "damagetaken",
    "buff",
    "protect_buff",
    "protect_cast",
};

local queue = {}
local protected = {}

local player_name = ""
local realm_name = ""
local player_config = nil
local rules = nil
local active_rules = nil

function battle_bot_add_rule( section, new_rule)
    new_rule["enabled"] = true
    new_rule["section"] = section

    local found = false

    local announce  = ""
    for index, rule in pairs(rules) do
        if( 
            rule['spellid'] == new_rule['spellid'] 
            and rule['section'] == new_rule['section']
        ) then
            rules[index] = new_rule
            announce = string.format(
                EPGP_BB_REPLACED_RULE
                , battle_bot_get_rule_as_string(rule)
                , battle_bot_get_rule_as_string(new_rule)
            )
            found = true
            break
        end
    end
    
    if( not found ) then
        table.insert(rules, new_rule)
        announce = string.format(
            EPGP_BB_CREATED_RULE
            , battle_bot_get_rule_as_string(new_rule)
        )
    end

    battle_bot_smart_announce(announce)
    battle_bot_make_active_rules()
end


function battle_bot_smart_announce(announce)
    if( announce ~= "" ) then
        if( UnitInRaid('player') == nil ) then
            print(announce)
        else
            SendChatMessage(announce, 'RAID')
        end
    end
end


function battle_bot_protect_handler( cmd, tail )
    -- /ebb protect by buff|cast spell_id
    local action, spell_id = string.match( tail, '^by%s+(%a+)%s+(%d+)$' )
    if( action == 'cast' ) then
        local item = {
            ['spellid'] = spell_id
        }
        battle_bot_add_rule( 'protect_cast', item )
    elseif( action == 'buff' ) then
        local item = {
            ['spellid'] = spell_id
        }
        battle_bot_add_rule( 'protect_buff', item )
    else
        EBB:HelpHandler();
    end
end


function battle_bot_add_handler( cmd, tail )                           
    local value, currency, action_base, action_ext, actor, tail = string.match( tail, '^(-?%d+)%s+(%ap)%s+for%s+(%a+)%s+(%w+)%s*(%w*)%s*(.*)$' )
    if( 
        action_base == 'damagetaken' 
        or action_base == 'death'
        or action_base == 'buff'
    ) then
        if( action_ext == 'by' ) then
            local item = {
                ['spellid'] = tonumber(actor),
                ['value'] = tonumber(value),
                ['currency'] = string.upper(currency),
            }

            -- buff min stacks
            if( action_base == 'buff' ) then
                item["stacks"] = tonumber(tail)
                if( item["stacks"] == nil ) then
                    item["stacks"] = 1
                end
            end

            -- min taken damage
            if( action_base == 'damagetaken' ) then
                item["damage_amount"] = tonumber(tail)
                if( item["damage_amount"] == nil ) then
                    item["damage_amount"] = 0
                end
            end
            
            battle_bot_add_rule( action_base, item )
        else
            EBB:HelpHandler();
        end
    elseif( 
        action_base == 'interrupt' 
        or action_base == 'dispel'
    ) then
        local spell_id = tonumber(action_ext)
        if( spell_id ~= nil ) then
            local item = {
                ['spellid'] = spell_id,
                ['value'] = tonumber(value),
                ['currency'] = string.upper(currency),
            }
            
            battle_bot_add_rule( action_base, item )
        else
            EBB:HelpHandler();
        end
    else
        EBB:HelpHandler();
    end
end


function battle_bot_get_rule_as_string( rule )
    local result

    local section = rule['section']
    if( section == 'death' ) then
        result = string.format(
            EPGP_BB_RULE_DEATH_BY_PH
            , rule["value"]
            , rule["currency"]
            , (GetSpellLink(rule["spellid"]))
        )
    elseif( section == 'damagetaken' ) then
        if( rule['damage_amount'] > 0 ) then
            result = string.format(
                EPGP_BB_RULE_DAMAGE_TAKEN_AMOUNT_BY_PH
                , rule["value"]
                , rule["currency"]
                , EBB:FormatNumber(rule["damage_amount"])
                , (GetSpellLink(rule["spellid"]))
            )
        else
            result = string.format(
                EPGP_BB_RULE_DAMAGE_TAKEN_BY_PH
                , rule["value"]
                , rule["currency"]
                , (GetSpellLink(rule["spellid"]))
            )
        end
    elseif( section == 'interrupt' ) then
        result = string.format(
            EPGP_BB_RULE_INTERRUPT_PH
            , rule["value"]
            , rule["currency"]
            , (GetSpellLink(rule["spellid"]))
        )
    elseif( section == 'dispel' ) then
        result = string.format(
            EPGP_BB_RULE_DISPEL_PH
            , rule["value"]
            , rule["currency"]
            , (GetSpellLink(rule["spellid"]))
        )
    elseif( section == 'protect_cast' ) then
        result = string.format(
            EPGP_BB_RULE_PROTECT_CAST_PH
            , (GetSpellLink(rule["spellid"]))
        )
    elseif( section == 'protect_buff' ) then
        result = string.format(
            EPGP_BB_RULE_PROTECT_BUFF_PH
            , (GetSpellLink(rule["spellid"]))
        )
    elseif( section == 'buff' ) then
        if( rule['stacks'] > 1 ) then
            result = string.format(
                EPGP_BB_RULE_BUFF_STACKS_BY_PH
                , rule["value"]
                , rule["currency"]
                , rule["stacks"]
                , (GetSpellLink(rule["spellid"]))
            )
        else
            result = string.format(
                EPGP_BB_RULE_BUFF_BY_PH
                , rule["value"]
                , rule["currency"]
                , (GetSpellLink(rule["spellid"]))
            )
        end
    else
        print( "Don't know how to stringify rule: "..section.."\n" )
        result = 'Unknown'
    end
    
    return result
end


function battle_bot_get_rules_text()
    local result = {}
    local counter = 1

    for _, item in pairs(rules) do
        local newitem = {
            ["enabled"] = item["enabled"],
            ["enabled_text"] = EPGP_BB_STATE[item["enabled"]],
            ["counter"] = counter,
            ["rule"] = battle_bot_get_rule_as_string(item),
            
        }

        table.insert( result, newitem )
        
        counter = counter + 1
    end
    
    return result
end


function battle_bot_list_handler( cmd, tail )
    local rules = battle_bot_get_rules_text()
    for _, rule in pairs(rules) do
        print(
            string.format(
                EPGP_BB_RULE_PH
                , rule["counter"]
                , rule["enabled_text"]
                , rule["rule"]
            )
        )
    end
end


function battle_bot_announce_handler( cmd, tail )
    local channel = string.lower(tail);
    
    if not( 
        channel == 'raid'
        or channel == 'guild'
        or channel == 'say'
        or channel == 'party'
    ) then
        if( UnitInRaid('player') == nil ) then
            channel = 'say'
        else
            channel = 'raid'
        end
    end
    
    local rules = battle_bot_get_rules_text(EPGP_BB_ACTIVE_RULE_PH)
   
    SendChatMessage(EPGP_BB_ACTIVE_RULES_HEADER, channel)
    for _, rule in pairs(rules) do
        if( rule["enabled"] ) then
            SendChatMessage( 
                "    "..rule["rule"]
                , channel
            )
        end
    end
end


function battle_bot_del_handler( cmd, tail )
    local old_tail = tail
    tail = tonumber(tail)
    if( rules[tail] ~= nil ) then
        local rule = table.remove(rules, tail)
        battle_bot_smart_announce(string.format(
            EPGP_BB_RULE_DELETED
            , battle_bot_get_rule_as_string(rule)
        ))
        battle_bot_make_active_rules()
    else
        print(string.format(EPGP_BB_RULE_NOT_FOUND, old_tail))
    end
    battle_bot_list_handler();
end


function battle_bot_enable_handler( cmd, tail )
    local old_tail = tail
    tail = tonumber(tail)
    if( rules[tail] ~= nil ) then
        rules[tail]["enabled"] = true
        battle_bot_smart_announce(string.format(
            EPGP_BB_RULE_ENABLED
            , battle_bot_get_rule_as_string(rules[tail])
        ))
        battle_bot_make_active_rules()
    else
        print(string.format(EPGP_BB_RULE_NOT_FOUND, old_tail))
        battle_bot_list_handler();
    end
end


function battle_bot_disable_handler( cmd, tail )
    local old_tail = tail
    tail = tonumber(tail)
    if( rules[tail] ~= nil ) then
        rules[tail]["enabled"] = false
        battle_bot_smart_announce(string.format(
            EPGP_BB_RULE_DISABLED
            , battle_bot_get_rule_as_string(rules[tail])
        ))
        battle_bot_make_active_rules()
    else
        print(string.format(EPGP_BB_RULE_NOT_FOUND, old_tail))
        battle_bot_list_handler();
    end
end

function battle_bot_turn_on_handler()
    player_config["enabled"] = true
    if( player_config["autologging"] ) then
        if( LoggingCombat(true) ) then
            battle_bot_smart_announce(EPGP_BB_LOGGING_ENABLED)
        else
            print("Error enabling logging")
        end
    end
    battle_bot_smart_announce(EPGP_BB_ADDON_ENABLED)
end


function battle_bot_turn_off_handler()
    player_config["enabled"] = false
    if( player_config["autologging"] ) then
        if( not LoggingCombat(false) ) then
            battle_bot_smart_announce(EPGP_BB_LOGGING_DISABLED);
        else
            print("Error disabling logging")
        end
    end
    battle_bot_smart_announce(EPGP_BB_ADDON_DISABLED)
end

function battle_bot_autologging_handler(cmd,tail)
    if( string.lower(tail) == 'on' ) then
        player_config['autologging'] = true
        print(EPGP_BB_AUTOLOGGING_ENABLED)
    elseif( string.lower(tail) == 'off' ) then
        player_config['autologging'] = false
        print(EPGP_BB_AUTOLOGGING_DISABLED)
    else
        EBB:HelpHandler()
    end
end

function battle_bot_reset_handler()
    player_config["rules"] = {}
    print(EPGP_BB_CONFIG_RESET)
end

function battle_bot_status_handler()
    print(string.format(EPGP_BB_STATUS_VERSION, EBB.version))
    print(string.format(EPGP_BB_STATUS_STATUS, EPGP_BB_STATE[player_config["enabled"]]))
    print(string.format(EPGP_BB_STATUS_AUTOLOGGING, EPGP_BB_STATE[player_config["autologging"]]))
    print(string.format(EPGP_BB_STATUS_RULES, table.getn(rules)))
end

function battle_bot_combatlog_parser(...)

   
end


-- legacy, this function should be removed in release or two
function battle_bot_make_active_rules()
    active_rules = {
        ["damagetaken"] = {},
        ["buff"] = {},
        ["death"] = {},
        ["dispel"] = {},
        ["interrupt"] = {},

        ["protect_buff"] = {},
        ["protect_cast"] = {},
    }
    for _, rule in pairs(rules) do
    
        local section = rule["section"]

        -- legacy migrating
        if( rule['value'] == nil and rule['gp_value'] ~= nil ) then
            rule['value'] = tonumber(rule['gp_value'])
            rule['gp_value'] = nil
        end
        
        if( rule['currency'] == nil ) then
            rule['currency'] = 'GP'
        end
        
        if( rule['value'] ~= tonumber(rule['value'])) then
            rule['value'] = tonumber(rule['value'])
        end
        
        if( rule['spellid'] ~= tonumber(rule['spellid'])) then
            rule['spellid'] = tonumber(rule['spellid'])
        end

        if( 
            rule['section'] == 'damagetaken' 
            and rule['damage_amount'] == nil 
        ) then
            rule['damage_amount'] = 0
        end
        
        rule['type'] = nil
        -- end of legacy migration
    
        if( rule["enabled"] ) then
            active_rules[section][rule["spellid"]] = rule
        end
    end
end

-- legacy, this function should be removed in release or two
function battle_bot_import_legacy_rules()
    for _, key in pairs(config_keys) do
        if( config[key] ~= nil ) then
            for _, rule in pairs( config[key] ) do
                table.insert(player_config["rules"], rule)
            end
            config[key] = nil
        end
    end
end


function battle_bot_check_config_keys()
    if( player_config["enabled"] == nil ) then
        player_config["enabled"] = true
    end
    if( player_config["autologging"] == nil ) then
        player_config["autologging"] = true
    end
    if( player_config["rules"] == nil ) then
        player_config["rules"] = {}
        battle_bot_import_legacy_rules()
    end    
end


----------------------- ACE section --------------------------------------------

EBB.version = GetAddOnMetadata("epgp_battle_bot", "Version")
EBB.slash_handlers = {
    status      = battle_bot_status_handler,
    autologging = battle_bot_autologging_handler,
    list        = battle_bot_list_handler,
    add         = battle_bot_add_handler,
    protect     = battle_bot_protect_handler,
    del         = battle_bot_del_handler,
    enable      = battle_bot_enable_handler,
    disable     = battle_bot_disable_handler,
    reset       = battle_bot_reset_handler,
    announce    = battle_bot_announce_handler,
    on          = battle_bot_turn_on_handler,
    off         = battle_bot_turn_off_handler,
}


function EBB:OnInitialize()
    EBB:Print("Initialise enabled")
    EBB:RegisterEvent("VARIABLES_LOADED")
    EBB:RegisterEvent("PLAYER_REGEN_DISABLED")
    EBB:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
    EBB:RegisterChatCommand("ebb", "SlashHandler")
end

function EBB:OnEnable()
    EBB:Print("EBB enabled")
end

function EBB:OnDisable()
    EBB:Print("EBB disabled")
end


function EBB:VARIABLES_LOADED()
    realm_name = SelectedRealmName()
    player_name = UnitName("player")
    EBB:ValidateConfig()
end

function EBB:PLAYER_REGEN_DISABLED()
    protected = {}
end

function EBB:COMBAT_LOG_EVENT_UNFILTERED(...)
    if( not player_config["enabled"] ) then
        return
    end

    local arg = {...}
    local timestamp, event, _, src_guid, src_name, src_flags, src_raid_flags, dst_guid, dst_name, dst_flags, dst_raid_flags, spell_id, spell_name, spell_school = ... -- 14 items
    
    local rule_target = dst_name
    local rule_target_guid = dst_guid
    local active_rule = nil
    
    if( event == "SPELL_AURA_APPLIED" ) then
        local rule = active_rules["buff"][spell_id]
        local buff_rule = active_rules["protect_buff"][spell_id]
        
        if( rule ~= nil and rule["stacks"] == 1 ) then
            active_rule = rule
        elseif( buff_rule ~= nil  and dst_name ~= nil ) then
            protected[dst_name] = true
        end        
    elseif( event == "SPELL_AURA_APPLIED_DOSE" ) then
        local rule = active_rules["buff"][spell_id]
   
        if( rule ~= nil and rule["stacks"] > 1 ) then
            local stacks = tonumber(arg[16])
            if( stacks ~= nil and stacks >= rule["stacks"] ) then
                active_rule = rule
            end
        end
    elseif( event == "SPELL_AURA_REMOVED" ) then 
        local buff_rule = active_rules["protect_buff"][spell_id]
        
        if( buff_rule ~= nil and dst_name ~= nil ) then
            protected[dst_name] = nil
        end        
    
    elseif( event == "SPELL_AURA_BROKEN" ) then 
        local buff_rule = active_rules["protect_buff"][spell_id]
        
        if( buff_rule ~= nil  and dst_name ~= nil ) then
            protected[dst_name] = nil
        end        
    elseif( event == "SPELL_DAMAGE" ) then
        local death_rule = active_rules["death"][spell_id]
        local damagetaken_rule = active_rules["damagetaken"][spell_id]
        local damage_amount, damage_overkill = tonumber(arg[15]), tonumber(arg[16]);
        
        -- Not yet necessary
        -- local damage_resisted = tonumber(arg[18]);
        -- local damage_blocked = tonumber(arg[19]);
        -- local damage_absorbed = tonumber(arg[20]);
        
        local damage_total = damage_amount + damage_overkill;
        if( damage_overkill == 0 ) then -- precise kill
            damage_total = damage_total + 1
        end
        
        if( death_rule ~= nil and damage_overkill > -1 ) then   -- death by spell
            active_rule = death_rule
        elseif( 
            damagetaken_rule ~= nil                                 -- got damagetaken rule
            and damage_total > damagetaken_rule['damage_amount']    -- passed threshold
            and ( 
                protected[dst_name] == nil  -- not protected
                or damage_overkill > -1     -- or died
            )
        ) then 
            active_rule = damagetaken_rule
        end
    elseif( event == "SPELL_CAST_START" ) then 
        local cast_rule = active_rules["protect_cast"][spell_id]
        
        if( cast_rule ~= nil  and src_name ~= nil ) then
            protected[src_name] = true
        end        
    elseif( event == "SPELL_CAST_SUCCESS" ) then 
        local cast_rule = active_rules["protect_cast"][spell_id]
        
        if( cast_rule ~= nil  and src_name ~= nil ) then
            protected[src_name] = nil
        end        
    
    elseif( event == "SPELL_CAST_FAILED" ) then 
        local cast_rule = active_rules["protect_cast"][spell_id]
        
        if( cast_rule ~= nil  and src_name ~= nil ) then
            protected[src_name] = nil
        end
    elseif( event == "SPELL_INTERRUPT" ) then 
        local target_spell = tonumber(arg[15])
        local rule = active_rules["interrupt"][target_spell]
        
        if( rule ~= nil and src_name ~= nil ) then
            active_rule, rule_target, rule_target_guid = rule, src_name, src_guid
        end
    elseif( 
        event == "SPELL_DISPEL" 
        or event == "SPELL_STOLEN" 
    ) then 
        local target_spell = tonumber(arg[15])
        local rule = active_rules["dispel"][target_spell]
        
        if( rule ~= nil and src_name ~= nil ) then
            active_rule, rule_target, rule_target_guid = rule, src_name, src_guid
        end
    elseif( event == "UNIT_DIED" ) then
        if( dst_name ~= nil ) then
            protected[dst_name] = nil
        end
    end

    -- push to queue, should check that we are in guild
    if( 
        active_rule ~= nil 
        and rule_target ~= nil 
        and (string.find(rule_target_guid, 'Player')) == 1
    ) then
        table.insert(queue, {
            ["player"] = rule_target,
            ["rule"] = active_rule
        })
    end

    -- proceed the queue
    if( 
        table.getn(queue) > 0 
        and GS:IsCurrentState()
    ) then 
        local item = table.remove(queue, 1);
        local rule = item['rule']
        local reason, value, currency, player = battle_bot_get_rule_as_string(rule), rule['value'], rule['currency'], item['player']
        
        if( currency == 'GP' ) then
            if( EPGP:CanIncGPBy(reason, value) ) then
                EPGP:IncGPBy( player, reason, value )
            else
                print("Unabe to charge GP", player, reason, value) -- debugging
            end
        else
            if( EPGP:CanIncEPBy(reason, value) ) then
                EPGP:IncEPBy( player, reason, value )
            else
                print("Unabe to charge EP", player, reason, value) -- debugging
            end
        end
    end
end

function EBB:ValidateConfig()
    if( config == nil ) then
        config = {}
    end

    if( config[realm_name] == nil ) then
        config[realm_name] = {}
    end

    if( config[realm_name][player_name] == nil ) then
        config[realm_name][player_name] = {}
    end
    
    player_config = config[realm_name][player_name];

    battle_bot_check_config_keys()

    rules = player_config["rules"]
    
    battle_bot_make_active_rules()
end


function EBB:SlashHandler( msg )
    msg = string.lower(msg)
    
    cmd, tail = string.match( msg, '^%s*(%a+)%s*(.*)$');
    
    if( self.slash_handlers[cmd] ) then
        self.slash_handlers[cmd](cmd, tail)
    else
        self:HelpHandler();
    end
end

function EBB:HelpHandler()
    for _, line in pairs(EPGP_BB_HELP) do
        print(line:format(EBB.version))
    end
end

-- http://lua-users.org/wiki/FormattingNumbers
function EBB:FormatNumber(amount)
    local formatted = amount

    while true do  
        formatted, k = string.gsub(formatted, "^(-?%d+)(%d%d%d)", '%1,%2')
        if (k==0) then
            break
        end
    end
    return formatted
end
